core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -979,6 +1004,21 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         }
     }
 
+    private void verifyPrimary() {
+        if (shardRouting.primary() == false) {
+            // must use exception that is not ignore be replication logic. See TransportActions.isShardNotAvailableException
+            throw new UnsupportedOperationException("shard is not a primary " + shardRouting);
+        }
+    }
+
+    private void verifyReplicationTarget() {
+        final IndexShardState state = state();
+        if (shardRouting.primary() && shardRouting.isRelocationTarget() == false && state != IndexShardState.RELOCATED) {
```

Comment:
OK I see the point. Thinking about this some more though, I think the following variation of the condition will make it clearer what we try to prevent:

```
if (shardRouting.primary() && shardRouting.active() && state != IndexShardState.RELOCATED) {
    // must use exception that is not ignored by replication logic. See TransportActions.isShardNotAvailableException
    throw new UnsupportedOperationException("active primary shard cannot be a replication target before relocation handoff " + shardRouting + ", state is [" + state + "]");
}
```

It precisely states that we cannot replicate into an active primary as long as it hasn't been relocated.